src/tools/crawler/crawl.rs

```diff
@@ -10,70 +11,170 @@ use serde::Deserialize;
+    /// Typically 51235.
+    /// Omitted if the peer is configured as a validator or a private peer.
+    pub port: Option<Port>,
+
+    /// The public key of the ECDSA key pair used by this peer to sign RTXP messages.
+    pub public_key: String,
+
+    /// Indicating whether the TCP connection to the peer is incoming or outgoing.
+    ///
+    /// The value is "in" or "out".
     #[serde(rename = "type")]
-    pub(super) connection_type: String,
-    pub(super) uptime: u32,
-    pub(super) version: String,
+    pub connection_type: String,
+
+    /// The number of seconds the server has been connected to this peer.
+    #[serde(rename = "uptime")]
+    pub connection_uptime: u32,
+
+    /// The rippled version number the peer reports to be using.
+    pub version: String,
+}
+
+impl fmt::Display for Peer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let ip = self.ip.clone().unwrap_or_default();
+        let port = self.port.clone().unwrap_or_default();
+        let public_key = &self.public_key;
+        let uptime = &self.connection_uptime;
+
+        writeln!(f, "[{uptime:05}]: {ip}:{port}\t\t\t{public_key}")?;
+        writeln!(
+            f,
+            "[{:3}]: \t\t\t\tversion: {:20} \t\tshards: {:?} ledgers: {:?}",
+            self.connection_type, self.version, self.complete_shards, self.complete_ledgers
+        )
+    }
 }
 
 impl Peer {
     /// Returns port number for the peer.
-    pub(super) fn port(&self) -> Option<u16> {
+    pub fn port(&self) -> Option<u16> {
         self.port.as_ref().and_then(|p| match p {
             Port::Number(n) => Some(*n),
             Port::String(s) => s.parse().ok(),
         })
     }
 }
 
-#[derive(Deserialize)]
-pub(super) struct CrawlResponse {
-    pub(super) overlay: Overlay,
-    pub(super) server: Server,
+#[derive(Debug, Deserialize)]
+pub struct CrawlResponse {
+    /// Information about the peer servers currently connected to this one,
+    /// similar to the response from the peers method.
+    #[serde(rename = "overlay")]
+    pub peerlist: Overlay,
+
+    /// Information about this server.
+    pub server: Server,
 }
 
-#[derive(Deserialize)]
-pub(super) struct Overlay {
-    pub(super) active: Vec<Peer>,
+impl fmt::Display for CrawlResponse {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, " server: {}", self.server)?;
+        writeln!(f, " peerlist:\n{}", self.peerlist)
+    }
 }
 
-#[allow(dead_code)]
-#[derive(Deserialize)]
-pub(super) struct Server {
-    pub(super) build_version: String,
-    pub(super) server_state: String,
-    pub(super) uptime: u32,
+#[derive(Debug, Deserialize)]
+pub struct Overlay {
+    pub active: Vec<Peer>,
 }
 
+impl fmt::Display for Overlay {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // Sort the peers here by the IP address.
+        let mut peers = self.active.clone();
+        peers.sort_by(|a, b| {
+            let a_ip = a.ip.clone().unwrap_or_default();
```

Comment:
Just wondering if cloning is here a must and something else could be used like:
`a.ip.map(|a| a.cmp(&b.ip.unwrap_or_default())).unwrap_or(Ordering::Less)`

Reply:
Ah, missed this one.

I like the idea, but since our `Option` contains a `String`, every `unwrap` will move out that value from the `Option` - I haven't found any quick solution to make it simpler without a `clone`